ic-agent/src/agent/builder.rs

```diff
@@ -34,19 +36,36 @@ impl AgentBuilder {
     /// Set a Replica transport to talk to serve as the replica interface.
     pub fn with_transport<F: 'static + ReplicaV2Transport>(self, transport: F) -> Self {
         Self {
-            config: AgentConfig {
+            config: AgentConfigImpl {
                 transport: Some(Arc::new(transport)),
                 ..self.config
             },
         }
     }
 
-    /// Add a NonceFactory to this Agent. By default, no nonce is produced.
-    pub fn with_nonce_factory(self, nonce_factory: NonceFactory) -> Self {
-        AgentBuilder {
-            config: AgentConfig {
+    /// Add a NonceFactory to this AgentImpl. By default, no nonce is produced.
+    pub fn with_nonce_factory(self, nonce_factory: NonceFactory) -> AgentBuilderImpl<NonceFactory> {
+        AgentBuilderImpl {
+            config: AgentConfigImpl {
                 nonce_factory,
-                ..self.config
+                identity: self.config.identity,
+                ingress_expiry_duration: self.config.ingress_expiry_duration,
+                transport: self.config.transport,
+            },
+        }
+    }
+
+    /// Add a NonceFactory to this AgentImpl. By default, no nonce is produced.
+    pub fn with_nonce_generator<N1: NonceGenerator>(
+        self,
+        nonce_factory: N1,
+    ) -> AgentBuilderImpl<N1> {
+        AgentBuilderImpl {
+            config: AgentConfigImpl {
+                nonce_factory,
+                identity: self.config.identity,
+                ingress_expiry_duration: self.config.ingress_expiry_duration,
+                transport: self.config.transport,
```

Comment:
Could this be like so?
```suggestion
                ...self.config
```

Reply:
I wish. But since they're not the same type (`N` is different), it gets all mad.